File: 卡码/代码随想录/1/01-数组/Go/01-二分/LeetCode_0034.go
//34. 在排序数组中查找元素的第一个和最后一个位置
package main

import "fmt"

func searchRange(nums []int, target int) []int {
	left,right:=0,len(nums)-1
	for left<=right {
		middle:=(left+right)/2
		if nums[middle]==target {
			start,end:=middle-1,middle+1
			for start>=0&&nums[start]==target {
				start--
			}
			for end<=len(nums)-1&&nums[end]==target {
				end++
			}
			return []int{start+1,end-1}
		}else if target<nums[middle] {
			right=middle-1
		}else{
			left=middle+1
		}
	}
	return []int{-1,-1}
}

func main(){
	nums:=[]int{5,7,7,8,8,10}
	fmt.Println(searchRange(nums,8))
	fmt.Println(searchRange(nums,6))
	fmt.Println(searchRange(nums,0))
}

File: 灵神/basicPractice/Java/basicPractice/src/main/java/com/kurumi/leetCode/basicPractice/二分/LeetCode_0034.java
package com.kurumi.leetCode.basicPractice.二分;

/**
 * @Author kurumi
 * @Date 2026-07-03 下午2:12:18
 * @Description 34. 在排序数组中查找元素的第一个和最后一个位置
 */
public class LeetCode_0034 {
    static class Solution {
        private int binarySearch(int[] nums, int target) {
            int left = 0, right = nums.length - 1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (nums[mid] < target) left = mid + 1;
                else right = mid - 1;
            }
            return left;
        }

        public int[] searchRange(int[] nums, int target) {
            int start = binarySearch(nums, target);
            if (start == nums.length || nums[start] != target) return new int[] {-1, -1};
            int end = binarySearch(nums, target + 1) - 1;
            return new int[] {start, end};
        }
    }
}
